AWSClientRuntime/Sources/Middlewares/Sha256TreeHashMiddleware.swift

```diff
@@ -0,0 +1,118 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0.
+
+import AwsCommonRuntimeKit
+import ClientRuntime
+import AwsCCal
+
+public struct Sha256TreeHashMiddleware<OperationStackOutput: HttpResponseBinding,
+                                       OperationStackError: HttpResponseBinding>: Middleware {
+    public let id: String = "Sha256TreeHash"
+    
+    private let sha256TreeHashHeaderName = "X-Amz-Sha256-Tree-Hash"
+    
+    private let sha256HashHeaderName = "X-Amz-Content-Sha256"
+    
+    public init() {}
+    
+    public func handle<H>(context: Context,
+                          input: MInput,
+                          next: H) -> Result<MOutput, MError>
+    where H: Handler,
+          Self.MInput == H.Input,
+          Self.MOutput == H.Output,
+          Self.Context == H.Context,
+          Self.MError == H.MiddlewareError {
+              let request = input.build()
+              
+              switch request.body {
+              case .data(let data):
+                  guard let data = data else {
+                            return next.handle(context: context, input: input)
+                        }
+                  if !request.headers.exists(name: sha256HashHeaderName) {
+                      let base64Encoded = ByteBuffer(data: data).base64EncodedSha256()
+                      input.withHeader(name: sha256TreeHashHeaderName, value: base64Encoded)
+                  }
+              case .stream(let stream):
+                  let streamBytes = stream.toBytes()
+                  guard streamBytes.length > 0 else {
+                      return next.handle(context: context, input: input)
+                  }
+                  let (linearHash, treeHash) = computeHashes(bytes: streamBytes)
+                  if let treeHash = treeHash {
+                      input.withHeader(name: sha256TreeHashHeaderName, value: treeHash)
+                  }
+                  if let linearHash = linearHash {
+                      input.withHeader(name: sha256HashHeaderName, value: linearHash)
+                  }
```

Comment:
It seems like we want to only want to add both of these headers or none, right?  If so, does the following work?
```
guard let (linearHash, treeHash) = computeHashes(...) else{
   return next.handle(..)
}
input.withHeader(name: sha256TreeHashHeaderName, treeHash)
input.withHeader(name: sha256HashHeaderName, linearHash)

```

Reply:
this desnt compile. cant do a guard let for a tuple of optionals for some reason but yes you are correct, ill combine them into one if let